processor/processor.go

```diff
@@ -0,0 +1,114 @@
+package processor
+
+import (
+	"fmt"
+	"path"
+
+	"github.com/algorand/go-algorand/agreement"
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/ledger"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/rpcs"
+)
+
+// Processor is the block processor interface
+type Processor interface {
+	Process(cert *rpcs.EncodedBlockCert) error
+	SetHandler(handler func(block *ledgercore.ValidatedBlock) error)
+	NextRoundToProcess() uint64
+	Init(path string, genesis *bookkeeping.Genesis, genesisBlock *bookkeeping.Block) error
```

Comment:
Additionally, I think `MakeProcessor()` should accept only `ledger *ledger.Ledger` and `handler`. The rationale is easier testability. You will be able to pass an in-memory `Ledger` instead of creating a temporary directory on disk (pass `true` to `ledger.OpenLedger()`).

Reply:
https://en.wikipedia.org/wiki/Dependency_injection

Reply:
I'm not sure `MakeProcessor(ledger *ledger.Ledger, handler func(block *ledgercore.ValidatedBlock))` is what we want. This definition seems to be having block fetcher and addBlock to create a ledger instance. Shouldn't processor be owning the local ledger?

Reply:
I think we can have the best of both approaches with two constructors:
* `MakeBlockProcessor(path string, genesis ...)` - self contained when part of a pipeline. 
* `makeBlockProcessor(ledger *ledger.Ledger)` - easy to inject the ledger for tests.

Reply:
I think this approach works for testing right now but it's not the best. 

1. it forces tests to be white box
2. if you write tests for a different package that depend on this package, you will not be able to use an in-memory ledger
3. in general this approach is inferior for a reason unrelated to testing; imagine you need to construct an object that itself constructs 10 objects with 5 parameters each. Instead of passing 50 parameters to the main object, it's better to construct 10 smaller objects and pass all of them to the main object.

Reply:
MakeBlockProcessor(path string, genesis ...) is really creating a local ledger, maybe I can rename it to CreateLedger and make it a helper method in`util`? and use MakeBlockProcessor(ledger *ledger.Ledger, handler func(block *ledgercore.ValidatedBlock)) in `processor`?

Reply:
Maybe make `func MakeInitState(genesis *bookkeeping.Genesis, genesisBlock *bookkeeping.Block) ledgercore.InitState` in this package? Only if you are going to test it though. Otherwise, I think it can go in `cmd/algorand-indexer/daemon.go`.